tests/federation_room_join_partial_state_test.go

```diff
@@ -513,6 +563,128 @@ func TestPartialStateJoin(t *testing.T) {
 	})
 }
 
+// test reception of an event over federation during a resync
+// sends the given event to the homeserver under test, checks that a client can see it and checks
+// the state at the event
+func testReceiveEventDuringPartialStateJoin(
+	t *testing.T, deployment *docker.Deployment, alice *client.CSAPI, psjResult partialStateJoinResult, event *gomatrixserverlib.Event,
+) {
+	// send the event to the homeserver
+	psjResult.Server.MustSendTransaction(t, deployment, "hs1", []json.RawMessage{event.JSON()}, nil)
+
+	/* TODO: check that a lazy-loading sync can see the event. Currently this doesn't work, because /sync blocks.
+	 * https://github.com/matrix-org/synapse/issues/13146
+	alice.MustSyncUntil(t,
+		client.SyncReq{
+			Filter: buildLazyLoadingSyncFilter(nil),
+		},
+		client.SyncTimelineHasEventID(psjResult.ServerRoom.RoomID, event.EventID()),
+	)
+	*/
+
+	// still, Alice should be able to see the event with an /event request. We might have to try it a few times.
+	start := time.Now()
+	for {
+		if time.Since(start) > time.Second {
+			t.Fatalf("timeout waiting for received event to be visible")
+		}
+		res := alice.DoFunc(t, "GET", []string{"_matrix", "client", "r0", "rooms", psjResult.ServerRoom.RoomID, "event", event.EventID()})
```

Comment:
You can now use [WithRetryUntil](https://github.com/matrix-org/complement/blob/main/internal/client/client.go#L459) in your `DoFunc` so you can remove the `for` loop etc. It would look a bit like:
```go
res := alice.DoFunc(t, "GET", []string{"_matrix", "client", "r0", "rooms", psjResult.ServerRoom.RoomID, "event", event.EventID()},
client.WithRetryUntil(time.Second, func(res *http.Response) bool {
    if res.StatusCode == 200 {
        return true
    }
    eventResBody := client.ParseJSON(t, res)
    if res.StatusCode == 404 && gjson.GetBytes(eventResBody, "errcode").String() == "M_NOT_FOUND" {
        return false
    }
    t.Fatalf("GET /event failed with %d: %s", res.StatusCode, string(eventResBody))
    return false
})
```